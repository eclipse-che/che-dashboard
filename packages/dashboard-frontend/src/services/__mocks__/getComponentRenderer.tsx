/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

export default function <T extends Array<unknown>>(
  getComponent: (..._args: T) => React.ReactElement,
): {
  createSnapshot: (...args: T) => renderer.ReactTestRenderer;
  renderComponent: (...args: T) => {
    reRenderComponent: (..._args: T) => void;
  };
} {
  return {
    createSnapshot: (...args: T) => {
      return renderer.create(getComponent(...args));
    },
    renderComponent: (...args: T) => {
      const res = render(getComponent(...args));
      return {
        reRenderComponent: (..._args: T) => {
          res.rerender(getComponent(..._args));
        },
      };
    },
  };
}
