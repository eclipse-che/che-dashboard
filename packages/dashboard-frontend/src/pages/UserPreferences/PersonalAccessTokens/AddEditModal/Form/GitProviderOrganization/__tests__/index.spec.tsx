/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Form } from '@patternfly/react-core';
import userEvent from '@testing-library/user-event';
import React from 'react';

import getComponentRenderer, { screen } from '@/services/__mocks__/getComponentRenderer';

import { GitProviderOrganization } from '..';

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

const mockOnChange = jest.fn();

describe('GitProviderOrganization', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('snapshot w/o organization', () => {
    const snapshot = createSnapshot();
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test('snapshot with organization', () => {
    const snapshot = createSnapshot('user-organization');
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  it('should handle a correct organization', async () => {
    const organization = 'user-organization';
    renderComponent();

    expect(mockOnChange).not.toHaveBeenCalled();

    const input = screen.getByRole('textbox');
    await userEvent.click(input);
    await userEvent.paste(organization);

    expect(mockOnChange).toHaveBeenCalledWith(organization, true);
    expect(screen.queryByText('This field is required.')).toBeFalsy();
    expect(screen.queryByText(/^The Provider Organization is too long./)).toBeFalsy();
  });

  it('should handle a too long organization value', async () => {
    // make it long enough to be invalid
    const organization = 'user-organization'.repeat(20);
    renderComponent();

    expect(mockOnChange).not.toHaveBeenCalled();

    const input = screen.getByRole('textbox');
    await userEvent.click(input);
    await userEvent.paste(organization);

    expect(mockOnChange).toHaveBeenCalledWith(organization, false);
    expect(screen.queryByText('This field is required.')).toBeFalsy();
    expect(screen.queryByText(/^The Provider Organization is too long./)).toBeTruthy();
  });

  it('should handle an empty value', async () => {
    const organization = 'user-organization';
    renderComponent(organization);

    expect(mockOnChange).not.toHaveBeenCalled();

    const input = screen.getByRole('textbox');
    await userEvent.clear(input);

    expect(mockOnChange).toHaveBeenCalledWith('', false);
    expect(screen.queryByText('This field is required.')).toBeTruthy();
    expect(screen.queryByText(/^The Provider Organization is too long./)).toBeFalsy();
  });
});

function getComponent(providerOrganization?: string): React.ReactElement {
  return (
    <Form>
      <GitProviderOrganization
        providerOrganization={providerOrganization}
        onChange={mockOnChange}
      />
    </Form>
  );
}
