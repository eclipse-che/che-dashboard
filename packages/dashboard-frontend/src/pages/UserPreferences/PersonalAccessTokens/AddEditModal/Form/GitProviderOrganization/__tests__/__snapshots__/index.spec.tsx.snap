// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitProviderOrganization snapshot w/o organization 1`] = `
<form
  className="pf-c-form"
  noValidate={true}
>
  <div
    className="pf-c-form__group"
  >
    <div
      className="pf-c-form__group-label"
    >
      <label
        className="pf-c-form__label"
        htmlFor="git-provider-organization-label"
      >
        <span
          className="pf-c-form__label-text"
        >
          Git Provider Organization
        </span>
        <span
          aria-hidden="true"
          className="pf-c-form__label-required"
        >
           
          *
        </span>
      </label>
       
    </div>
    <div
      className="pf-c-form__group-control"
    >
      <input
        aria-describedby="git-provider-organization-label"
        aria-invalid={false}
        aria-label="Git Provider Organization"
        className="pf-c-form-control"
        data-ouia-component-id="OUIA-Generated-TextInputBase-1"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe={true}
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Enter a Git Provider Organization"
        required={false}
        type="text"
        value=""
      />
      
    </div>
  </div>
</form>
`;

exports[`GitProviderOrganization snapshot with organization 1`] = `
<form
  className="pf-c-form"
  noValidate={true}
>
  <div
    className="pf-c-form__group"
  >
    <div
      className="pf-c-form__group-label"
    >
      <label
        className="pf-c-form__label"
        htmlFor="git-provider-organization-label"
      >
        <span
          className="pf-c-form__label-text"
        >
          Git Provider Organization
        </span>
        <span
          aria-hidden="true"
          className="pf-c-form__label-required"
        >
           
          *
        </span>
      </label>
       
    </div>
    <div
      className="pf-c-form__group-control"
    >
      <input
        aria-describedby="git-provider-organization-label"
        aria-invalid={false}
        aria-label="Git Provider Organization"
        className="pf-c-form-control"
        data-ouia-component-id="OUIA-Generated-TextInputBase-2"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe={true}
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Enter a Git Provider Organization"
        required={false}
        type="text"
        value="user-organization"
      />
      
    </div>
  </div>
</form>
`;
